### PR TITLE
[CES-86] - Cosmos account module

### PR DESCRIPTION
**Pull request by christian-calabrese**
# Azure Cosmos DB Account Module

This module provisions an Azure Cosmos DB account with configurable features and settings. It's designed to be flexible and support various deployment scenarios while adhering to best practices for security and performance.

## Key Features

1. **Geo-Replication**
2. **Private endpoints**
3. **Encryption - Customer managed key**
4. **Consistency Configuration**
5. **System and user assigned identities supported**

## Usage
Refer to the example configuration and input variables for detailed usage instructions.

### CODE REVIEW

File: infra/modules/azure_cosmos_account/cosmos.tf [hunk 1]
```diff
@@ -0,0 +1,46 @@
+resource "azurerm_cosmosdb_account" "this" {
+  name                          = "${local.app_name_prefix}-cosno-${var.environment.instance_number}"
+  location                      = var.environment.location
+  resource_group_name           = var.resource_group_name
+  offer_type                    = "Standard"
```
Comment (gunzip) [inline infra/modules/azure_cosmos_account/cosmos.tf:5]: can you add a comment about why we use 'standard'?

Comment (Krusty93) [reply infra/modules/azure_cosmos_account/cosmos.tf:5]: It's the only available value as far as I know

File: infra/modules/azure_cosmos_account/cosmos.tf [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    max_staleness_prefix    = var.consistency_policy.consistency_level == "BoundedStaleness" ? var.consistency_policy.max_staleness_prefix : null
+  }
+
+  # As suggested by technology https://pagopa.atlassian.net/wiki/spaces/DEVOPS/pages/500039691/Cosmos+DB#backup-policy
+  backup {
+    type = "Continuous"
+  }
+
+  identity {
+    type         = var.customer_managed_key.enabled ? "SystemAssigned, UserAssigned" : "SystemAssigned"
```
Comment (Krusty93) [inline infra/modules/azure_cosmos_account/cosmos.tf:41]: I can't think of a case where we'd need to assign a system-assigned managed identity only. Might it be worth to remove it?

File: infra/modules/azure_cosmos_account/locals.tf [hunk 3]
```diff
@@ -0,0 +1,11 @@
+locals {
```
Comment (Krusty93) [inline infra/modules/azure_cosmos_account/locals.tf:1]: Sync with @mamu0 as he's refactoring this part

File: infra/modules/azure_cosmos_account/variables.tf [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  }
+}
+
+variable "force_public_network_access_enabled" {
+  type        = bool
+  description = "(Optional) Whether the Storage Account permits public network access or not. Defaults to false."
+  default     = false
+}
+
+variable "automatic_failover_enabled" {
```
Comment (Krusty93) [inline infra/modules/azure_cosmos_account/variables.tf:84]: I would think about forcing this value to true

File: infra/modules/azure_cosmos_account/examples/complete/main.tf [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  ]
+
+  customer_managed_key = {
+    enabled                   = true
+    user_assigned_identity_id = azurerm_user_assigned_identity.example.id
+    key_vault_key_id          = "https://${local.environment.prefix}-${local.environment.env_short}-kv.vault.azure.net/keys/my-kv-key"
+  }
+
+  force_public_network_access_enabled = false
+  automatic_failover_enabled          = true
```
[the remark this answers is not in the document]
Comment (christian-calabrese) [reply infra/modules/azure_cosmos_account/examples/complete/main.tf:42]: As correctly suggested by @Krusty93, automatic failover has been forced to true

File: infra/modules/azure_cosmos_account/examples/complete/main.tf [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  source = "../../"
+
+  environment         = local.environment
+  resource_group_name = azurerm_resource_group.example.name
+  tags                = local.tags
+
+  subnet_pep_id                        = data.azurerm_subnet.pep.id
+  private_dns_zone_resource_group_name = "${local.environment.prefix}-${local.environment.env_short}-rg-common"
+
+  primary_geo_location = {
```
[the remark this answers is not in the document]
Comment (christian-calabrese) [reply infra/modules/azure_cosmos_account/examples/complete/main.tf:22]: What do you think about adding an example with the minimum set of required variables?

File: infra/modules/azure_cosmos_account/examples/complete/main.tf [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    user_assigned_identity_id = azurerm_user_assigned_identity.example.id
+    key_vault_key_id          = "https://${local.environment.prefix}-${local.environment.env_short}-kv.vault.azure.net/keys/my-kv-key"
+  }
+
+  force_public_network_access_enabled = false
+  automatic_failover_enabled          = true
+
+  consistency_policy = {
+    consistency_level       = "BoundedStaleness"
+    max_interval_in_seconds = 300
```
[the remark this answers is not in the document]
Comment (christian-calabrese) [reply infra/modules/azure_cosmos_account/examples/complete/main.tf:46]: No, they are pretty random. Discussing with @Krusty93, we thought that the consistency level is very dependent on how the application uses the database.
I think there's not a one size fits all configuration unfortunately

[the remark this answers is not in the document]
Comment (christian-calabrese) [reply infra/modules/azure_cosmos_account/examples/complete/main.tf:46]: If you agree, we can integrate the description with this consideration

File: infra/modules/azure_cosmos_account/cosmos.tf (hunk 2 above)
Comment (christian-calabrese) [reply infra/modules/azure_cosmos_account/cosmos.tf]: True, unless we don't use an encryption key in the same account it's probably useless. Changed!

File: infra/modules/azure_cosmos_account/examples/complete/main.tf (hunk 7 above)
[the remark this answers is not in the document]
Comment (gunzip) [reply infra/modules/azure_cosmos_account/examples/complete/main.tf:45]: we should have some presets (s, m, l) as teams aren't aware abput the meaning of these values and shouldn't

File: infra/modules/azure_cosmos_account/examples/complete/main.tf (hunk 6 above)
[the remark this answers is not in the document]
Comment (gunzip) [reply infra/modules/azure_cosmos_account/examples/complete/main.tf:22]: that would be great!

File: infra/modules/azure_cosmos_account/examples/complete/main.tf (hunk 7 above)
[the remark this answers is not in the document]
Comment (christian-calabrese) [reply infra/modules/azure_cosmos_account/examples/complete/main.tf:45]: An initial set of presets have been implemented, but can be enlarged in the future. The values for bounded staleness have been retrieved from the only configured cosmos account we have in our tenant

File: infra/modules/azure_cosmos_account/cosmos.tf [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+  consistency_policy {
+    consistency_level = var.consistency_policy.consistency_level
+
+    # Only apply these fields if the consistency_level is "BoundedStaleness"
+    max_interval_in_seconds = var.consistency_policy.consistency_level == "BoundedStaleness" ? var.consistency_policy.max_interval_in_seconds : null
+    max_staleness_prefix    = var.consistency_policy.consistency_level == "BoundedStaleness" ? var.consistency_policy.max_staleness_prefix : null
+  }
+
+  # As suggested by technology https://pagopa.atlassian.net/wiki/spaces/DEVOPS/pages/500039691/Cosmos+DB#backup-policy
```
[the remark this answers is not in the document]
Comment (christian-calabrese) [reply infra/modules/azure_cosmos_account/cosmos.tf:35]: A variable has just been added to choose between the possible 7 or 30 days. It defaults to 30 days as of now

**Comment by gunzip** [comment]
lgtm, let's wait for @pagopa/engineering-team-devex review

File: infra/modules/azure_cosmos_account/variables.tf [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+  }
+}
+
+variable "force_public_network_access_enabled" {
+  type        = bool
+  description = "(Optional) Whether the Storage Account permits public network access or not. Defaults to false."
+  default     = false
+}
+
+variable "consistency_policy" {
```
Comment (mamu0) [inline infra/modules/azure_cosmos_account/variables.tf:84]: Is it necessary for this variable to be highly configurable? Perhaps we could identify a recurring preset that better fits our use cases

Comment (christian-calabrese) [reply infra/modules/azure_cosmos_account/variables.tf:84]: Actually, after [Danilo's comment](https://github.com/pagopa/dx/pull/130#discussion_r1791742751) some presets have been implemented (Default, HighConsistency, HighPerformance, BalancedStaleness)

However, by setting the consistency_preset to "Custom", it is possible to configure the consistency settings in detail (but still not needed if you are not a cosmos ninja 😄 🥷 )

**Review by mamu0 — APPROVED** [review]
LGTM

File: infra/modules/azure_cosmos_account/cosmos.tf [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    # Only apply these fields if the consistency_level is "BoundedStaleness"
+    max_interval_in_seconds = local.final_consistency_policy.consistency_level == "BoundedStaleness" ? local.final_consistency_policy.max_interval_in_seconds : null
+    max_staleness_prefix    = local.final_consistency_policy.consistency_level == "BoundedStaleness" ? local.final_consistency_policy.max_staleness_prefix : null
+  }
+
+  # As suggested by technology https://pagopa.atlassian.net/wiki/spaces/DEVOPS/pages/500039691/Cosmos+DB#backup-policy
+  backup {
+    type = "Continuous"
+    tier = "Continuous${var.point_in_time_restore_retention_days}Days"
```
Comment (Krusty93) [inline infra/modules/azure_cosmos_account/cosmos.tf:38]: What do you think to put `30` here?

**Review by Krusty93 — APPROVED** [review]

**Pull request merged**